enhancements/network/on-prem-bgp-networking.md

```diff
@@ -0,0 +1,447 @@
+
+### API Extensions
+
+None.
+
+### Implementation Details/Notes/Constraints
+
+#### MetalLB
+
+MetalLB is a load-balancer implementation for bare metal Kubernetes clusters, using standard routing protocols and backed by FRR when using BGP as a protocol.
+
+For now, we recommend separating the MetalLB deployment from the workers where Ingress Controller runs.
+This is because the MetalLB deployment is configured to use the same BGP peers as the ones used by our implementation.
+The [MetalLB documentation](https://metallb.universe.tf/configuration/_advanced_bgp_configuration/#limiting-peers-to-certain-nodes) explains how to decide where to run peers on certain nodes by using the node-selectors attribute of peers in the Peers configuration. The semantics of these selectors are the same as those used elsewhere in Kubernetes, so refer to the [labels documentation](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels).
+
+#### Customer requirements
+
+With this new implementation, the customer needs to have a BGP infrastructure in place to be able to deploy and use the OpenShift cluster (which is why we will never make it the default implementation).
+
+Here is a list of requirements that the customer needs to have in place:
+
+- A BGP router (or multiple) that can peer with the BGP peers on the control plane nodes.
+- ICMP redirects are disabled on the router(s) that will peer with the control plane nodes. This is needed to ensure that the router(s) does not send an ICMP redirect message when traffic destined to the VIPs is received on a node where the VIP is not configured. This is needed to ensure that the traffic is not black-holed in the network.
+
+We will need to document the requirements and the steps to configure the BGP infrastructure, as well as providing as much as possible some validation tools to help the customer to configure the BGP infrastructure correctly.
+
+#### Networking infrastructure maintenance
+
+Router(s) that will peer with the control plane nodes need to be maintained by the customer, but in case of a failure, it's possible that the BGP configuration has to change.
+Therefore, we need to provide a way for the customer to update the BGP configuration of the control plane nodes.
+The customer would update the BGP Configuration via the OpenShift API, and baremetal-runtimecfg which would monitor the Infrastructure object would update the configuration of the control plane nodes and reload FRR.
+
+### Risks and Mitigations
+
+- This design heavily relies on the infrastructure network that is not controlled by OpenShift. Documentation and validations will be provided
+  to ensure that prerequisites are met before a deployment can be done using BGP
+- Performance can be degraded if infrastructure network has routing problems, but routing issues would also involve a general outage, not just control plane VIPs
```

Comment:
can we detect this? (should we? or "not our problem"?)

Reply:
I'll add a note about this.